File: quantum1D/quantum_1D.m
function quantum_1D

Nstep = 200;      %总步数
X = round(linspace(-Nstep,Nstep,2*Nstep+1));  %粒子所有可能到达的位置(横轴)

psi = zeros(2,2*Nstep+1);
psi(:,Nstep+1) = 1/sqrt(2) * [1, 1j]';  % |ψ>的初态
c = 1/sqrt(2) * [1, 1; 1, -1]; % quantum coin转移矩阵

%> 固定步数Nstep, 观察最后的粒子分布
% %< 态演化过程
% for i = 1:Nstep
%     temp = c * psi; % c|ψ>
%     psi(1,:) = [0 temp(1,1:end-1)];
%     psi(2,:) = [temp(2,2:end) 0];
% end
% %< 观测粒子分布
% p = zeros(1,2*Nstep+1);
% for i = 1:2*Nstep+1
%     p(i) = dot(abs(psi(:,i)),abs(psi(:,i)));
% end
% %< 画图
% plot(X,p);


%> 观察粒子分布随时间(步数)的演化
tic;
p = zeros(Nstep,2*Nstep+1);
for i = 1:Nstep

    temp = c * psi; % c|ψ>
    psi(1,:) = [0 temp(1,1:end-1)];
    psi(2,:) = [temp(2,2:end) 0];
    % psi(1,2:end) = temp(1,1:end-1);
    % psi(2,1:end-1) = temp(2,2:end);

    for k = 1:2*Nstep+1
        p(i,k) = dot(abs(psi(:,k)),abs(psi(:,k)));
    end

end
toc




save("C:\Users\JAH\Desktop\组\random walk\quantum1D\quantum1D.mat","p","X","Nstep");
fprintf("数据已导出");

File: quantum1D/quantum_1D.py
from os import X_OK
import numpy as np
import matplotlib.pyplot as plt
import matplotlib as mpl
import scipy.io as sio
from matplotlib import animation
from matplotlib.animation import FuncAnimation
from matplotlib import colors
from mpl_toolkits.mplot3d import Axes3D

data = sio.loadmat(r'C:\Users\JAH\Desktop\组\random walk\quantum1D\quantum1D.mat')

p = np.array(data["p"])
Nstep = int(data["Nstep"])
X = np.array(data["X"])[0]

fig, ax1 = plt.subplots()

#> 观察粒子分布随时间(步数)的演化

def update(i):

    ax1.clear()    #清除上一帧的axes对象
    ax1.plot(X,p[i],linewidth=0.2)

    ax1.set_title("1D quantum random walk")
    ax1.set_xlabel('X')
    ax1.set_ylabel('p')

    # ax1.set_xlim(-Nstep,Nstep)
    ax1.set_ylim(0,1)

    return ax1


anim = FuncAnimation(fig, update, frames=np.arange(0, Nstep), interval=50, blit=False, repeat=False)
# plt.show()
anim.save(r'C:\\Users\\JAH\\Desktop\\组\\random walk\\quantum1D\\1D quantum probability distribution.gif')